flink-python/src/main/java/org/apache/flink/streaming/api/utils/LengthPrefixWrapperSerializer.java

```diff
@@ -0,0 +1,197 @@
+        return dataSerializer.copy(from, reuse);
+    }
+
+    @Override
+    public int getLength() {
+        return bytesSerializer.getLength();
+    }
+
+    @Override
+    public void serialize(T record, DataOutputView target) throws IOException {
+        dataSerializer.serialize(record, baosWrapper);
+        bytesSerializer.serialize(baos.toByteArray(), target);
+        baos.reset();
+    }
+
+    @Override
+    public T deserialize(DataInputView source) throws IOException {
+        byte[] bytes = bytesSerializer.deserialize(source);
+        bais.setBuffer(bytes, 0, bytes.length);
+        return dataSerializer.deserialize(baisWrapper);
+    }
+
+    @Override
+    public T deserialize(T reuse, DataInputView source) throws IOException {
+        byte[] bytes = bytesSerializer.deserialize(source);
+        bais.setBuffer(bytes, 0, bytes.length);
+        return dataSerializer.deserialize(reuse, baisWrapper);
+    }
+
+    @Override
+    public void copy(DataInputView source, DataOutputView target) throws IOException {
+        bytesSerializer.copy(source, target);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return obj.getClass().equals(this.getClass()) && bytesSerializer.equals(obj);
+    }
+
+    @Override
+    public int hashCode() {
+        return bytesSerializer.hashCode();
+    }
+
+    @Override
+    public TypeSerializerSnapshot<T> snapshotConfiguration() {
+        return new LengthPrefixWrapperSerializerSnapshot<>(
+                bytesSerializer.snapshotConfiguration(), dataSerializer.snapshotConfiguration());
+    }
+
+    /**
+     * {@link LengthPrefixWrapperSerializerSnapshot} wraps a serializer snapshot for byte[] and
+     * another for {@link T}.
+     */
+    public static class LengthPrefixWrapperSerializerSnapshot<T>
+            implements TypeSerializerSnapshot<T> {
+
+        private final TypeSerializerSnapshot<byte[]> bytesSerializerSnapshot;
+
+        private final TypeSerializerSnapshot<T> dataSerializerSnapshot;
+
+        public LengthPrefixWrapperSerializerSnapshot(
+                TypeSerializerSnapshot<byte[]> bytesSerializerSnapshot,
+                TypeSerializerSnapshot<T> dataSerializerSnapshot) {
+            this.bytesSerializerSnapshot = bytesSerializerSnapshot;
+            this.dataSerializerSnapshot = dataSerializerSnapshot;
+        }
+
+        @Override
+        public int getCurrentVersion() {
+            return dataSerializerSnapshot.getCurrentVersion();
+        }
+
+        @Override
+        public void writeSnapshot(DataOutputView out) throws IOException {
+            bytesSerializerSnapshot.writeSnapshot(out);
+            dataSerializerSnapshot.writeSnapshot(out);
+        }
+
+        @Override
+        public void readSnapshot(int readVersion, DataInputView in, ClassLoader userCodeClassLoader)
+                throws IOException {
+            bytesSerializerSnapshot.readSnapshot(readVersion, in, userCodeClassLoader);
```

Comment:
I guess we should use *TypeSerializerSnapshot.readVersionedSnapshot* here.

Could we add a test case for this class?